thor/observations/observations.py

```diff
@@ -0,0 +1,234 @@
+        # Quivr interface (add column -- add a column to a table)
+        # If add column is un-quivr like then I'd create a temporary table here
+        # unique_time_obscodes = unique_time_obscodes.add_column("state_id", qv.Int64Column(pa.array()))
+        unique_time_obscodes.insert(0, "state_id", unique_time_obscodes.index.values)
+
+        ### 4. Merge the unique state IDs back into the times_obscodes table
+        # Quivr interface (join -- create composite table schema from two tables)
+        # times_obscodes = times_obscodes.join(unique_time_obscodes, on=["time_mjd", "observatory_code"])
+
+        # Merge the unique state IDs back into the times_obscodes table
+        times_obscodes = pd.merge(
+            times_obscodes, unique_time_obscodes, on=["time_mjd", "observatory_code"]
+        )
+
+        ### 5. Merge the detections table with the times_obscodes table to get the state IDs for each detection
+        # Quivr interface (join -- create composite table schema from two tables)
+        # detections_states = detections.join(times_obscodes, on=["id", "exposure_id"], columns=["state_id"])
+
+        # Now merge the detections table with the times_obscodes table to get the state IDs for each detection
+        detections_states = pd.merge(
+            detections.to_dataframe(),
+            times_obscodes[["obs_id", "exposure_id", "state_id", "observatory_code"]],
+            left_on=["id", "exposure_id"],
+            right_on=["obs_id", "exposure_id"],
+        )
+
+        ### 6. Sort the detections by time and observatory code
+        # Quivr interface (sort -- sort a table by one or more columns)
+        # detections_states = detections_states.sort(by=["time_mjd", "observatory_code"], ascending=[True, True])
+        detections_states.sort_values(
+            by=["time.jd1", "time.jd2", "observatory_code"],
+            inplace=True,
+            ignore_index=True,
+        )
+
+        # In summary, with the proposed quivr interface, this function would look like:
+        # times_obscodes = detections.join(exposures, on="exposure_id", columns=["exposure_id", "observatory_code"])
+        # unique_times_obscodes = times_obscodes.drop_duplicates(subset=["time", "observatory_code"])
+        # unique_times_obscodes = unique_times_obscodes.add_column("state_id", qv.Int64Column(pa.array()))
+        # times_obscodes = times_obscodes.join(unique_times_obscodes, on=["time", "observatory_code"])
+        # detections_states = detections.join(times_obscodes, on=["id", "exposure_id"], columns=["state_id"])
+        # detections_states = detections_states.sort(by=["time", "observatory_code"], ascending=[True, True])
+        return cls.from_kwargs(
+            detections=PointSourceDetections.from_flat_dataframe(detections_states),
+            observatory_code=detections_states["observatory_code"],
+            state_id=detections_states["state_id"],
+        )
+
```

Comment:
I have rewritten this using only pyarrow table operations, will push it once its tested a bit more.